### PR TITLE
Add Python type annotations

### PR DESCRIPTION
**Pull request by zero323**
This PR adds type annotations for Python module, provided as stub files.

This should:

- Improve auto-completion performance in editors which support type hints.
- Optionally enable static checking through mypy or built-in tools (for example Pycharm tools).
- Provide additional way to document usage of certain methods, where arguments are required, despite providing default values.

### CODE REVIEW

**Comment by databricks-cla-assistant** [comment]
[CLA assistant check](https://cla.dev.databricks.com/delta-io/delta?pullRequest=305) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.databricks.com/delta-io/delta?pullRequest=305) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dev.databricks.com/check/delta-io/delta?pullRequest=305) it.</sub>

File: python/delta/testing/utils.pyi [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import unittest
```
Comment (zero323) [inline python/delta/testing/utils.pyi]: Depending on preferences test annotations could be dropped, as long as py.typed indicates partial annotations.

**Comment by tdas** [comment]
@zero323 my apologies for not having taken a look at this PR. I am unfamiliar with this aspect of python. can you tell me more about these annotations? Specifically
- How does it affect forward and backward compatibility?  
- Does these annotations always have to be in sync with the code? What happens if they are not in sync (e.g., we add a method in the python code but forget to add it to the annotation)?

**Comment by zero323** [comment]
> @zero323 my apologies for not having taken a look at this PR. 

No worries.

I am unfamiliar with this aspect of python. can you tell me more about these annotations? Specifically
> 
> * How does it affect forward and backward compatibility?

In case of stub files there is no practical on forward or backward compatibility (inline annotation, not proposed here, are not backward compatible beyond Python 3.5). Annotations are only used by external tools, and have no impact on actual execution.

> * Does these annotations always have to be in sync with the code? What happens if they are not in sync (e.g., we add a method in the python code but forget to add it to the annotation)?

Obviously this is the preferred to keep things consistent, as annotations which are not synced, provide misleading feedback. This might be a problem, if user use CI pipelines, which are aware of type annotations (type checkers and such). 

Other than that, there is no impact at all (might be in the future, if users couple things with emerging  tools like [`mypyc`](https://github.com/python/mypy/tree/master/mypyc)).

**Comment by zsxwing** [comment]
> Since I opened this PR, Spark started moving from stubs to inline hints. Should we consider the same here?

@zero323 Thanks for pointing out this. I think that's better. Could you update your PR using inline hints?

**Comment by zero323** [comment]
> Could you update your PR using inline hints?

I'll work on that, but it might take a while.

**Comment by zero323** [comment]
@zsxwing, could you be so kind and approve the workflow? TIA!

File: python/delta/tables.py [hunk 2]
```diff
@@ -870,12 +1016,17 @@ def addColumns(self, cols):
...
+        scalaSchema = jsparkSession.parseDataType(cols.json())
         self._jbuilder = self._jbuilder.addColumns(scalaSchema)
         return self
 
-    @since(1.0)
-    def partitionedBy(self, *cols):
+    @since(1.0)  # type: ignore[arg-type]
+    def partitionedBy(
+        self, *cols: Union[str, List[str], Tuple[str, ...]]
+    ) -> "DeltaTableBuilder":
```
[the remark this answers is not in the document]
Comment (scottsand-db) [reply python/delta/tables.py:1029]: @zero323 and @ueshin  what's the status of this discussion?
@ueshin are you okay with the current implementation?

**Comment by scottsand-db** [comment]
@zero323 this seems to be ready for another round of reviews after your latest commits.
I'll re-request reviews from @xinrong-databricks and @ueshin

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

File: python/delta/tests/test_deltatable.py [hunk 3]
```diff
@@ -368,8 +400,9 @@ def __verify_table_schema(self, tableName, schema, cols,
             fields.append(StructField(col, dataType, col in nullables, metadata))
         assert (StructType(fields) == schema)
         if len(properties) > 0:
-            tablePropertyMap = self.spark.sql("SHOW TBLPROPERTIES {}".format(tableName)) \
-                .rdd.collectAsMap()
+            tablePropertyMap: Dict[str, str] = (
+                self.spark.sql("SHOW TBLPROPERTIES {}".format(tableName))  # type: ignore
```
[the remark this answers is not in the document]
Comment (zero323) [reply python/delta/tests/test_deltatable.py]: Done (This was intentional choice, as this is just not verifiable usage pattern).

File: python/delta/tables.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zero323) [reply python/delta/tables.py:1029]: Let me just add overload here, and leave discussion about variance handling for later, after we settle on the right approach upstream.

**Review by xinrong-databricks — APPROVED** [review]

**Comment by xinrong-databricks** [comment]
LGTM if tests passed. Thanks!

**Comment by scottsand-db** [comment]
Hi @zero323 thanks for all the work on this PR. Almost there! this PR is failing python tests since commit `417dc53` with error (below). Can you please take a look and fix the failing tests?

```
python/delta/tests/test_deltatable.py:404: error: Invalid self argument "RDD[Row]" to attribute function "collectAsMap" with type "Callable[[RDD[Tuple[K, V]]], Dict[K, V]]"  [misc]
python/delta/tests/test_deltatable.py:404: error: Incompatible types in assignment (expression has type "Dict[<nothing>, <nothing>]", variable has type "Dict[str, str]")  [assignment]
Found 2 errors in 1 file (checked 12 source files)
Deleted /home/runner/.ivy2/cache/io.delta 
### Executing cmd: /home/runner/work/delta/delta/build/sbt clean publishM2
### Executing cmd: /home/runner/work/delta/delta/dev/lint-python
##### Running mypy tests #####
### Executing cmd: mypy --config-file /home/runner/work/delta/delta/python/mypy.ini /home/runner/work/delta/delta/python/delta
Traceback (most recent call last):
  File "python/run-tests.py", line 175, in <module>
    run_mypy_tests(root_dir)
  File "python/run-tests.py", line 113, in run_mypy_tests
    ], stream_output=True)
  File "python/run-tests.py", line 83, in run_cmd
    raise Exception("Non-zero exitcode: %s" % (exit_code))
Exception: Non-zero exitcode: 1
##### Running SBT tests #####
##### Running Python tests #####
Calling script %s python/run-tests.py
Traceback (most recent call last):
  File "run-tests.py", line 76, in <module>
    run_python_tests(root_dir)
  File "run-tests.py", line 34, in run_python_tests
    run_cmd(["python", python_test_script], stream_output=True)
  File "run-tests.py", line 46, in run_cmd
    raise Exception("Non-zero exitcode: %s" % (exit_code))
Exception: Non-zero exitcode: 1
Error: Process completed with exit code 1.
```

**Comment by zero323** [comment]
@scottsand-db I've seen this and I am investigating, but getting inconsistent behavior for ignores between local dev environment

```
➜  delta git:(annotations) git rev-parse HEAD
4ea8fff69a4631c91984abcc854a8fedcd26677c
➜  delta git:(annotations) mypy --version                                 
mypy 0.910
➜  delta git:(annotations) rm -Rf .mypy_cache                             
➜  delta git:(annotations) sha256sum python/delta/tests/test_deltatable.py
d34acbe3ad488c65b1734fadf03e1732161f2c8a27e26b8ece72bb39d4030670  python/delta/tests/test_deltatable.py
➜  delta git:(annotations) mypy --no-incremental --config-file python/mypy.ini python/delta 
Success: no issues found in 12 source files
```

and CI setup

```
dev@7fd042106847:~/delta$ git rev-parse HEAD
4ea8fff69a4631c91984abcc854a8fedcd26677c
dev@7fd042106847:~/delta$ pipenv run mypy --version
mypy 0.910
dev@7fd042106847:~/delta$ sha256sum python/delta/tests/test_deltatable.py 
d34acbe3ad488c65b1734fadf03e1732161f2c8a27e26b8ece72bb39d4030670  python/delta/tests/test_deltatable.py
dev@7fd042106847:~/delta$ rm -Rf .mypy_cache/
dev@7fd042106847:~/delta$ pipenv run mypy --no-incremental --config-file python/mypy.ini python/delta
python/delta/tests/test_deltatable.py:404: error: Invalid self argument "RDD[Row]" to attribute function "collectAsMap" with type "Callable[[RDD[Tuple[K, V]]], Dict[K, V]]"  [misc]
python/delta/tests/test_deltatable.py:404: error: Incompatible types in assignment (expression has type "Dict[<nothing>, <nothing>]", variable has type "Dict[str, str]")  [assignment]
Found 2 errors in 1 file (checked 12 source files)
```
I'll ping you all, once I figure out what causes the issue.

**Comment by zero323** [comment]
This seems to be unrelated to PR (already present on https://github.com/delta-io/delta/commit/5571d95966e3b03c4699dcfa1a5cd86bb16447dc)

```
 ImportError: cannot import name 'InstallCommand' from 'pipenv.vendor.pip_shims.shims' (/usr/local/lib/python3.8/dist-packages/pipenv/vendor/pip_shims/shims.py)
/usr/share/python-wheels/requests-2.22.0-py2.py3-none-any.whl/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.6) or chardet (3.0.4) doesn't match a supported version!
```

**Comment by zsxwing** [comment]
@zero323 I'm fixing it in https://github.com/delta-io/delta/pull/821

**Comment by zsxwing** [comment]
@zero323 you can merge master into your PR to fix the build.

**Comment by zero323** [comment]
Thanks @zsxwing!

**Comment by zero323** [comment]
I really cannot tell what causes difference in ignore handling between CI and local env, but both seem to be happy with the current position of  `type: ignore` (seems weird to me, TBH).

**Review by scottsand-db — APPROVED** [review]

**Comment by scottsand-db** [comment]
@zero323 do we need the empty file `python/delta/py.typed` ?

**Comment by zero323** [comment]
> @zero323 do we need the empty file `python/delta/py.typed` ?

Good catch @scottsand-db!  Yes, we definitely do. That's what makes package recognized as supporting type hints (PEP 561). Which reminded me that it should be mentioned in setup (added 267b6be)

**Comment by scottsand-db** [comment]
Hi @@zero323 , thanks for all of this work. I'm in the process of QAing and merging this ticket.

Can you please provide me with a tangible example of where + how this PR improves python typed annotations? and how you tested that?

I appreciate the help - not much of a python expert myself.

**Comment by zero323** [comment]
> Can you please provide me with a tangible example of where + how this PR improves python typed annotations? and how you tested that?
> 
> I appreciate the help - not much of a python expert myself.

A good starting point to understand the impact is to see corresponding Spark JIRA [ticket](https://issues.apache.org/jira/browse/SPARK-32681). To summarize, there three main aspects:

- Precise documentation. Expected input and output types are clearly defined. Furthermore, we can describe possible relations between arguments. For `delta`, the most obvious example are overloads of functions with counter-intuitive signatures. For example, take a look at `whenMatchedUpdate`.

  Because of this design decision

  https://github.com/delta-io/delta/blob/267b6be43b206924165baed3e51660f1379c28db/python/delta/tables.py#L714-L715

  The original signature

   https://github.com/delta-io/delta/blob/0f30f6f47761cb01043d84aad182f30cb59b5e31/python/delta/tables.py#L596

  is misleading, suggesting that is OK to omit `set`. In contrast, annotated code clearly indicates that `set` is mandatory:

  https://github.com/delta-io/delta/blob/267b6be43b206924165baed3e51660f1379c28db/python/delta/tables.py#L688-L704

- Ability to statically detect many code issues. 

   This is already visible in the delta codebase, where mypy can detect incorrect calls within tests intended to fail ‒ hence all the `ignores`, i.e.

   https://github.com/delta-io/delta/blob/267b6be43b206924165baed3e51660f1379c28db/python/delta/tests/test_deltatable.py#L107-L126


   We focus on mypy here, as the standard Python type checker, but depending on the environment (IDEs can come with their own type checkers providing different set of features) and type checker (there quite a few) your experience might be different.

- Finally, type hints enable better syntax completion. Again, specific behavior depends on your environment (@ueshin  and @xinrong-databricks  might be able to answer how this interacts with Databricks notebooks), but here is a little demo how experience changes in IPython (which depends on awesome [jedi](https://github.com/davidhalter/jedi) library by  [davidhalter/](https://github.com/davidhalter/) ‒ common choice for different projects) between 0f30f6f47761cb01043d84aad182f30cb59b5e31 and 267b6be43b206924165baed3e51660f1379c28db

  [asciicast](https://asciinema.org/a/TyWTbNNXRkk8h4YBHRTAPJ9L7)

**Comment by scottsand-db** [comment]
Thank you!

**Review by zsxwing — APPROVED** [review]
@zero323 Thanks a lot! The video is pretty cool!

**Comment by zero323** [comment]
Thanks all!

**Pull request closed without merging**